main/src/menu.ts

```diff
@@ -23,20 +23,39 @@ export const buildMenu = (mainWindow, loadInit) => {
     // for VIM as it sees escape keys.
     const normalizePath = (fileName) => fileName.split("\\").join("/")
 
-    const executeVimCommand = (command) => mainWindow.webContents.send("menu-item-click", command)
+    const ensureWindow = (currentWindow: BrowserWindow) => {
+        return !!currentWindow ? currentWindow : createWindow([], process.cwd())
```

Comment:
I'm wondering if this should return a `Promise`? In other words - I'm not sure if we will properly handle all of the commands we send if we are sending them immediately when the window is created. I believe we might need to wait for the window to be fully initialized before we send some of these commands.

An example of this would be `menu-item-click` and `open-files` - these are listened for in the `NeovimEditor` class, which might take a bit of time after opening to be initialized and start listening for those commands.

Reply:
@bryphe thats definitely the case the issue I encountered though is that the initialisation just triggers an event which then sets off other async processes but the createWindow function itself from what I can see is *sync*, I'm not entirely sure how to make it wait with an arbitrary timeout which is a messy/hacky solution, I though maybe of having the ipcMain process send an event when neovim is initialised and then responding to it in an event handler in the main process sort of the reverse of what we've been doing for the most part so outside in, not sure if thats possible

Reply:
Ah ok, luckily for us, there is a way we can make `createWindow` async - or at least return a `Promise` - to help us out here.

> I thought maybe of having the ipcMain process send an event when neovim is initialised and then responding to it in an event handler in the main process sort of the reverse of what we've been doing for the most part so outside in, not sure if thats possible

Yes, definitely - this is a good idea!

We already hook an event today to figure out when the initialization happens:
https://github.com/onivim/oni/blob/c01ae065c853e50471b7e4b17f4cadbb019fe7f7/main/src/main.ts#L141

What we could do is the following:
- From `createWindow`, we return a `Promise`
- In that `did-finish-load` handler (or our handler we add), we resolve the promise we returned
- We gate the command handlers on the `createWindow` promise (using `await`)

That should be enough to make things work. If the timing still isn't quite right, we could send an explicit event from the browser process -> to the main process letting it know that we're completely initialized.

As you mentioned, though, we might want to explicitly send a message up from the renderer process. We could do that somewhere here: https://github.com/onivim/oni/blob/c01ae065c853e50471b7e4b17f4cadbb019fe7f7/browser/src/index.tsx#L139

And have the main process listen for that event, in order to resolve the promise.

Hope that helps give you some ideas - let me know if you have any questions on the specifics. I think we can get this working though 👍